internal/cmd/marketplace/apply.go

```diff
@@ -0,0 +1,319 @@
+
+var (
+	errResWithoutName       = errors.New(`the required field "name" was not found in the resource`)
+	errNoValidFilesProvided = errors.New("no valid files were provided, see errors above")
+	errDuplicatedResName    = errors.New("some resources have duplicated name field")
+	errResNameNotAString    = errors.New(`the field "name" must be a string`)
+	errInvalidExtension     = errors.New("file has an invalid extension. Valid extensions are `.json`, `.yaml` and `.yml`")
+)
+
+func listFilesInDirRecursive(rootPath string) ([]string, error) {
+	var files []string
+	err := filepath.Walk(rootPath, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if !info.IsDir() {
+			files = append(files, path)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return files, nil
+}
+
+func buildPathsListFromDir(dirPath string) ([]string, error) {
+	filesPaths, err := listFilesInDirRecursive(dirPath)
+	if err != nil {
+		return nil, err
+	}
+	filePaths := []string{}
+	for _, path := range filesPaths {
+		switch filepath.Ext(path) {
+		case encoding.YamlExtension:
+			fallthrough
+		case encoding.YmlExtension:
+			fallthrough
+		case encoding.JSONExtension:
+			filePaths = append(filePaths, path)
+		default:
+			return nil, fmt.Errorf("%w: %s", errInvalidExtension, path)
+		}
+	}
+	return filePaths, nil
+}
+
+func buildApplyRequest(pathList []string) (*ApplyRequest, error) {
+	resources := []Item{}
+	resNameToFilePath := map[string]string{}
+	for _, filePath := range pathList {
+		content, err := os.ReadFile(filePath)
+		if err != nil {
+			return nil, err
+		}
+		var fileEncoding string
+		switch filepath.Ext(filePath) {
+		case encoding.YamlExtension:
+			fallthrough
+		case encoding.YmlExtension:
+			fileEncoding = YAML
+		case encoding.JSONExtension:
+			fileEncoding = JSON
+		default:
+			return nil, fmt.Errorf("%w: %s", errInvalidExtension, filePath)
+		}
+		marketplaceItem := &Item{}
+		err = encoding.UnmarshalData(content, fileEncoding, marketplaceItem)
+		if err != nil {
+			return nil, fmt.Errorf("errors in file %s: %w", filePath, err)
+		}
+		resName, err := retrieveAndValidateResName(*marketplaceItem, resNameToFilePath, filePath)
+		if err != nil {
+			return nil, err
+		}
+		resources = append(resources, *marketplaceItem)
+		resNameToFilePath[resName] = filePath
+	}
+	if len(resources) == 0 {
+		return nil, errNoValidFilesProvided
+	}
+	return &ApplyRequest{
+		Resources: resources,
+	}, nil
+}
+
+func retrieveAndValidateResName(res Item, resNameToFilePath map[string]string, filePath string) (string, error) {
+	resName, ok := res["name"]
+	if !ok {
+		return "", fmt.Errorf("%w: %s", errResWithoutName, filePath)
+	}
+	resNameStr, ok := resName.(string)
+	if !ok {
+		return "", fmt.Errorf("%w: %s", errResNameNotAString, filePath)
+	}
+	if _, ok := resNameToFilePath[resNameStr]; ok {
+		return "", fmt.Errorf("%w: %s", errDuplicatedResName, resName)
```

Comment:
Might it be better for the future to add specific error related infos directly into the error avoiding to wrap it?

Reply:
but this would prevent error matching with `errors.Is()` by callers and tests. This is why I preferred this approach. 
Another approach could be creating a struct that implements `Error` interface, but I thought it was too much for this use case

Reply:
I agree, for this case it may be overkill but I would prefer the struct way in general!

Reply:
I get your point, probably in a next development we can refactor it